### ISSUE TITLE
[BUG] Performance downgrade using OPTIMIZE (compaction) with Trino

### ISSUE THREAD

**Issue by Dearkano** [delta-io/delta#1151]
### Describe the problem
We were happy to see that OSS delta lake supported compaction, so we upgraded and tried to optimize our current tables. However, we observed the performance downgrade after compaction.

We use Starburst Enterprise (Trino) to query the delta lake on the top of S3, and here's our performance test using TPC-DS, and the chart below shows the size and object numbers of each table. 
![image](https://user-images.githubusercontent.com/24840232/170155351-1b38c435-c809-4577-980f-acfbccef160c.png)

After compaction, each table's files were merged into one file, one example below.
![Screen Shot 2022-05-24 at 5 53 28 PM](https://user-images.githubusercontent.com/24840232/170155462-c084fa48-dc38-4495-98b6-1803df8ca10b.png)

Then we did the benchmark test and the comparsion, the below is the result of the query elapsed time, the left (blue) is before the compaction, and the right (red) is after.
![Screen Shot 2022-05-24 at 5 41 30 PM](https://user-images.githubusercontent.com/24840232/170154418-59077074-23df-43fe-8898-8a10677e966c.png)

The samiliar downgraded performance was also observed on our data (we have large partitioned tables, and test the benchmarks on select with different filters and complex join queries)

### Environment information

* Delta Lake version: 1.2.1
* Spark version: 3.2.1
* Scala version: 2.12
* Starburst/Trino version: 370-e.3

### Willingness to contribute

The Delta Lake Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the Delta Lake code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the Delta Lake community.
- [x] No. I cannot contribute a bug fix at this time.

**Comment by vkorukanti** [comment]
@Dearkano Thanks for reporting the issue. Is it possible to share the Trino query execution profile _before optimize_ and _after optimize_ for one of the TPCDS queries (preferably Q21 which seems to have the worst degradation)?

**Comment by Dearkano** [comment]
@vkorukanti  sure, I will do the test and upload them later

**Comment by Dearkano** [comment]
@vkorukanti  @scottsand-db 
Hi, after we dive into the queries, we found that the root cause is that:
After the compaction, the parallelism of Trino is significantly reduced, this is expected since Trino has only one file (in the TPC-DS case) to deal with in each table after compaction. So CPU time after compaction is only 1/2 or even 1/10 compared to before. 
But in the end, the queries take a longer elapsed time, with consuming fewer resources.

Below is the comparison of the elapsed time, it should be similar to my original post, but this time, we ingested two copies of TPC-DS and then optimized one of them.
![Screen Shot 2022-05-27 at 6 09 05 PM](https://user-images.githubusercontent.com/24840232/170803837-01f9588d-9341-4c07-92fa-e7a2ade62efa.png)

And this one is the comparison of the CPU time, we can see that the compaction saves the CPU significantly.
![Screen Shot 2022-05-27 at 6 09 59 PM](https://user-images.githubusercontent.com/24840232/170803868-4a49c257-64b3-4cb1-b848-f02e32cd8402.png)

Take one query as an example, the query text is:
```
select count(*) 
from ((select distinct c_last_name, c_first_name, d_date
       from store_sales, date_dim, customer
       where store_sales.ss_sold_date_sk = date_dim.d_date_sk
         and store_sales.ss_customer_sk = customer.c_customer_sk
         and d_month_seq between 1197 and 1197+11)
       except
      (select distinct c_last_name, c_first_name, d_date
       from catalog_sales, date_dim, customer
       where catalog_sales.cs_sold_date_sk = date_dim.d_date_sk
         and catalog_sales.cs_bill_customer_sk = customer.c_customer_sk
         and d_month_seq between 1197 and 1197+11)
       except
      (select distinct c_last_name, c_first_name, d_date
       from web_sales, date_dim, customer
       where web_sales.ws_sold_date_sk = date_dim.d_date_sk
         and web_sales.ws_bill_customer_sk = customer.c_customer_sk
         and d_month_seq between 1197 and 1197+11)
) cool_cust

```

This is the execution details of the query before compaction.
![Screen Shot 2022-05-27 at 6 12 41 PM](https://user-images.githubusercontent.com/24840232/170803942-97cf30cf-d8ca-44e0-810e-415f422932b8.png)
![Screen Shot 2022-05-27 at 6 19 23 PM](https://user-images.githubusercontent.com/24840232/170804144-5152152e-fc86-4552-b37c-edc3327c01f6.png)

This is the execution details of the query after compaction.
![Screen Shot 2022-05-27 at 6 12 53 PM](https://user-images.githubusercontent.com/24840232/170803954-ec446be6-349f-4df3-af5e-e645f38efeb5.png)
![Screen Shot 2022-05-27 at 6 19 43 PM](https://user-images.githubusercontent.com/24840232/170804155-7beb34c3-2f9b-4189-a40f-f6e5acb3a0e9.png)


Since we do not have cases where in each partition, there are tens of GBs of data (so there will be tens of files after compaction), we do not have benchmarks on cases like this. 

In conclusion, the compaction will help Trino save resources while this may sacrifice the parallelism of Trino and make queries slower if the query resource is sufficient.

**Comment by vkorukanti** [comment]
@Dearkano From the screenshots you posted, it looks like the number of stages is different? Is the query plan different when running on a compacted data?

**Comment by Dearkano** [comment]
@vkorukanti Thank you for the reply!
I have not tried max-split-size, do you mean that we can use this config to increase the parallelism?
I attach the query plans here. From my side, the plan is indeed different just like the stage shows. There are 10 fragments in the plan with compaction, and 16 fragments in the plan without compaction. I'm not very familiar with the query plan so I'm not sure if there is any join optimization you mentioned.

[Query Plan - Before Compaction.txt](https://github.com/delta-io/delta/files/9049449/Query.Plan.-.Before.Compaction.txt)
[Query Plan - After Compaction.txt](https://github.com/delta-io/delta/files/9049452/Query.Plan.-.After.Compaction.txt)

**Comment by zsxwing** [comment]
Closing the stale issue. Feel free to reopen it if the above suggestion doesn't work

**Comment by ryadav-uptycs** [comment]
@Dearkano  is this fixed for you . I am also facing same issue , that on compacted files queries performance is slow or same vs querying uncompacted data .
